File: training/vue/src/de/sots/cellarsandcentaurs/adapter/persistence/CreatureRepository.ts
import { CreatureEntity as Entity } from './CreatureEntity'
import { Repository } from './Repository'

export class CreatureRepository extends Repository<Entity> {
  async findByPrefix(prefix: string): Promise<Entity[]> {
    const all: Entity[] = await this.findAll()
    return all.filter((entity) => entity.id.startsWith(prefix))
  }
}

File: sample-projects/ruby/lib/de/sots/cellars_and_centaurs/adapter/persistence/creature_repository.rb
# frozen_string_literal: true

require_relative '../../version'
require_relative 'repository'
require_relative 'creature_entity'

module De::Sots::CellarsAndCentaurs::Adapter::Persistence
  Entity = CreatureEntity

  class CreatureRepository < Repository
    def find_all
      to_a
    end

    def find_by_type(type)
      select { |entity| entity.type == type }
    end

    private

    def key_of(entity)
      raise ArgumentError, 'not an entity' unless entity.is_a?(Entity)

      entity.id
    end
  end
end
